compiler/sem/semstmts.nim

```diff
@@ -1080,14 +1080,26 @@ proc semNormalizedConst(c: PContext, n: PNode): PNode =
       c.p.owner.kind != skMacro:
     typFlags.incl taProcContextIsNotMacro
 
-  def = semConstExpr(c, def)
-
-  if def.kind != nkNilLit:
-    # xxx: this seems wrong, instead of guarding on nil, maybe it should be
-    #      inferred from `typ`? then we still do the rest of the checks?
-    if c.matchedConcept != nil:
-      typFlags.incl taConcept
-    typ = typeAllowedOrError(typ, skConst, c, def, typFlags)
+  if c.matchedConcept != nil:
+    typFlags.incl taConcept
+
+  # semantically analyse the expression first, so that it can be rejected
+  # early based on the type before trying to evaluate it
+  def = semExprWithType(c, def)
+  typ = typeAllowedOrError(typ, skConst, c, def, typFlags)
+
+  # XXX: ``fitNode`` is never used here. That doesn't seem correct, as the
```

Comment:
I agree that looks to be an oversight, I think in many cases it'd to unnoticed, but it does seem like a bug.

Reply:
This comment was an oversight by me. When splitting the evaluation part from `semConstExpr` and replacing its usage in `semNormalizedConst`, I also added the call to `semExprWithType` in order to mirror the behavior of `semConstExpr` also does.

Since a call to `fitNode` has to happen *after* a `semExprX` and there was none following the call to `semConstExpr`, I made the (incorrect) assumption that `fitNode` wasn't used. A few lines above however, the rhs is both analysed *and* fitted. In other words, there was no issue (except maybe that the already semantically analysed expression was analysed again).